aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,1719 @@
+    }
+
+    fun assert_delegation_pool_exists(pool_address: address) {
+        assert!(delegation_pool_exists(pool_address), error::invalid_argument(EDELEGATION_POOL_DOES_NOT_EXIST));
+    }
+
+    /// Retrieves the shared resource account owning the stake pool in order
+    /// to forward a stake-management operation to this underlying pool.
+    fun retrieve_stake_pool_signer(pool: &DelegationPool): signer {
+        account::create_signer_with_capability(&pool.stake_pool_signer_cap)
+    }
+
+    /// Get the address of delegation pool reference `pool`.
+    fun get_pool_address(pool: &DelegationPool): address {
+        account::get_signer_capability_address(&pool.stake_pool_signer_cap)
+    }
+
+    #[view]
+    /// Return the index of current observed lockup cycle on delegation pool `pool_address`.
+    /// This represents the key into `inactive_shares` of the `pending_inactive` shares pool.
+    public fun observed_lockup_cycle(pool_address: address): u64 acquires DelegationPool {
+        assert_delegation_pool_exists(pool_address);
+        borrow_global<DelegationPool>(pool_address).observed_lockup_cycle
+    }
+
+    /// Allows an owner to change the operator of the underlying stake pool.
+    public entry fun set_operator(owner: &signer, new_operator: address) acquires DelegationPoolOwnership, DelegationPool {
+        let pool_address = get_owned_pool_address(signer::address_of(owner));
+        // synchronize delegation and stake pools before any user operation
+        synchronize_delegation_pool(pool_address);
+        stake::set_operator(&retrieve_stake_pool_signer(borrow_global<DelegationPool>(pool_address)), new_operator);
+    }
+
+    /// Allows an owner to change the delegated voter of the underlying stake pool.
+    public entry fun set_delegated_voter(owner: &signer, new_voter: address) acquires DelegationPoolOwnership, DelegationPool {
+        let pool_address = get_owned_pool_address(signer::address_of(owner));
+        // synchronize delegation and stake pools before any user operation
+        synchronize_delegation_pool(pool_address);
+        stake::set_delegated_voter(&retrieve_stake_pool_signer(borrow_global<DelegationPool>(pool_address)), new_voter);
+    }
+
+    #[view]
+    /// Return the fee to be charged for an `add_stake` operation of `amount` on pool at `pool_address`.
+    /// If the validator produces rewards this epoch, added stake goes directly to `pending_active` and
+    /// does not earn rewards. However, all shares within a pool_u64 appreciate uniformly, when this epoch ends:
+    /// - either added shares are still `pending_active` and steal from rewards of existing `active` stake
+    /// - or have moved to `pending_inactive` and get full rewards (they displaced `active` stake at `unlock`)
+    /// Therefore, should charge delegator the maximum amount it would unfairly earn only this epoch.
+    public fun get_add_stake_fee(pool_address: address, amount: u64): u64 {
```

Comment:
One nuance I realized about this "fee" is that users don't technically "lose" this amount. Example:
1. Pool has 900 coins currently. Current number of shares is also 900 for simplicity.
2. I add 100 coins. Let's say rewards rate per epoch is 1%. So the fee for me is 1 coin. The pool now has 1000 coins but I only have 99 shares. Total number of shares is 999.
3. At the end of the epoch, the pool now has 1000 + 900 * 0.01 = 1009 coins.
4. I still have 99 shares but those shares are now worth 1009 / 999 * 99 = ~99.99 coins. This is almost the same as the original amount I added (100)

So technically at the end of the current epoch, I "regain" the fee. I would regain 100% of it if the validator performs at 100% in the current epoch. I recommend we avoid the term "fee" but may be phrase it as "buy in at the settlement price of current epoch" or something similar.